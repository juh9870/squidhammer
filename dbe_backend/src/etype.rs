use crate::etype::default::DefaultEValue;
use crate::etype::econst::ETypeConst;
use crate::etype::eitem::EItemInfo;
use crate::etype::eobject::EObject;
use crate::json_utils::{json_expected, json_kind, JsonValue};
use crate::m_try;
use crate::registry::{EObjectType, ETypesRegistry};
use crate::value::id::{EListId, EMapId, ETypeId};
use crate::value::EValue;
use miette::{bail, miette, Context};
use ordered_float::OrderedFloat;
use serde::{Deserialize, Serialize};
use smallvec::{smallvec, SmallVec};
use std::borrow::Cow;
use std::collections::BTreeMap;
use std::ops::Deref;
use std::sync::LazyLock;
use strum::EnumIs;
use ustr::Ustr;
use utils::whatever_ref::{WhateverRef, WhateverRefMap};

pub mod conversion;
pub mod default;
pub mod econst;
pub mod eenum;
pub mod eitem;
pub mod eobject;
pub mod estruct;
pub mod property;
pub mod title;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, EnumIs)]
pub enum EDataType {
    /// Primitive boolean type
    Boolean,
    /// Primitive numeric type
    Number,
    /// Primitive string type
    String,
    /// Inline object, enum, or list type
    Object {
        ident: ETypeId,
    },
    /// Primitive constant type
    Const {
        value: ETypeConst,
    },
    List {
        id: EListId,
    },
    Map {
        id: EMapId,
    },
    /// Unknown type. Cannot be named by the kdl type system, but can be used
    /// in graphs
    Unknown,
}

impl EDataType {
    pub fn default_value(&self, reg: &ETypesRegistry) -> DefaultEValue {
        match self {
            EDataType::Boolean => EValue::Boolean { value: false },
            EDataType::Number => EValue::Number { value: 0.0.into() },
            EDataType::String => EValue::String {
                value: Default::default(),
            },
            EDataType::Object { ident } => return reg.default_value_inner(ident),
            EDataType::Const { value } => value.default_value(),
            EDataType::List { id } => EValue::List {
                id: *id,
                values: vec![],
            },
            EDataType::Map { id } => EValue::Map {
                id: *id,
                values: Default::default(),
            },
            EDataType::Unknown => EValue::Null,
        }
        .into()
    }

    pub const fn null() -> EDataType {
        EDataType::Const {
            value: ETypeConst::Null,
        }
    }

    /// Returns the name of the type for debugging purposes
    pub fn name(&self) -> Cow<'_, str> {
        match self {
            EDataType::Boolean => "boolean".into(),
            EDataType::Number => "number".into(),
            EDataType::String => "string".into(),
            EDataType::Object { ident } => ident.to_string().into(),
            EDataType::Const { value } => value.to_string().into(),
            EDataType::List { id: ty } => ty.to_string().into(),
            EDataType::Map { id: ty } => ty.to_string().into(),
            EDataType::Unknown => "unknown".into(),
        }
    }

    /// Returns the human-readable title of the type
    pub fn title(&self, registry: &ETypesRegistry) -> String {
        match self {
            EDataType::Boolean | EDataType::Number | EDataType::String | EDataType::Unknown => {
                self.name().to_string()
            }
            EDataType::Object { ident } => registry.get_object(ident).map_or_else(
                || format!("Unknown object `{}`", ident),
                |data| data.title(registry),
            ),
            EDataType::Const { value } => value.to_string(),
            EDataType::List { id } => registry.get_list(id).map_or_else(
                || format!("Unknown list `{}`", id),
                |data| format!("List<{}>", data.value_type.title(registry)),
            ),
            EDataType::Map { id } => registry.get_map(id).map_or_else(
                || format!("Unknown map `{}`", id),
                |data| {
                    format!(
                        "Map<{}, {}>",
                        data.key_type.title(registry),
                        data.value_type.title(registry)
                    )
                },
            ),
        }
    }

    /// Returns the generic arguments names for this type
    pub fn generic_arguments_names<'a>(
        &self,
        registry: &'a ETypesRegistry,
    ) -> impl Deref<Target = [Ustr]> + 'a {
        enum Names<'a> {
            Ref(&'a [Ustr]),
            Map(WhateverRefMap<'a, EObjectType, [Ustr]>),
        }

        impl Deref for Names<'_> {
            type Target = [Ustr];
            fn deref(&self) -> &Self::Target {
                match self {
                    Names::Ref(r) => r,
                    Names::Map(r) => r.deref(),
                }
            }
        }

        match self {
            EDataType::Boolean
            | EDataType::Number
            | EDataType::String
            | EDataType::Const { .. }
            | EDataType::Unknown => Names::Ref(&[]),
            EDataType::Object { ident } => {
                let obj = registry.get_object(ident).expect("object should exist");
                Names::Map(WhateverRef::map(obj, |obj| obj.generic_arguments_names()))
            }
            EDataType::List { .. } => {
                static NAMES: LazyLock<[Ustr; 1]> = LazyLock::new(|| [Ustr::from("Item")]);
                Names::Ref(NAMES.deref())
            }
            EDataType::Map { .. } => {
                static NAMES: LazyLock<[Ustr; 2]> =
                    LazyLock::new(|| [Ustr::from("Key"), Ustr::from("Item")]);
                Names::Ref(NAMES.deref())
            }
        }
    }

    /// Returns the generic arguments values for this type
    pub fn generic_arguments_values<'a>(
        &self,
        registry: &'a ETypesRegistry,
    ) -> impl Deref<Target = [EItemInfo]> + 'a {
        enum Names<'a> {
            Nil,
            Vec(SmallVec<[EItemInfo; 2]>),
            Ref(WhateverRefMap<'a, EObjectType, [EItemInfo]>),
        }

        impl Deref for Names<'_> {
            type Target = [EItemInfo];
            fn deref(&self) -> &Self::Target {
                match self {
                    Names::Nil => &[],
                    Names::Vec(v) => v.as_slice(),
                    Names::Ref(r) => r.deref(),
                }
            }
        }

        match self {
            EDataType::Boolean
            | EDataType::Number
            | EDataType::String
            | EDataType::Const { .. }
            | EDataType::Unknown => Names::Nil,
            EDataType::Object { ident } => {
                let obj = registry.get_object(ident).expect("object should exist");
                Names::Ref(WhateverRef::map(obj, |obj| obj.generic_arguments_values()))
            }
            EDataType::List { id } => {
                let list = registry.get_list(id).expect("list should exist");
                Names::Vec(smallvec![EItemInfo::simple_type(list.value_type)])
            }
            EDataType::Map { id } => {
                let map = registry.get_map(id).expect("map should exist");
                Names::Vec(smallvec![
                    EItemInfo::simple_type(map.key_type),
                    EItemInfo::simple_type(map.value_type),
                ])
            }
        }
    }

    pub fn parse_json(
        &self,
        registry: &ETypesRegistry,
        data: &mut JsonValue,
        inline: bool,
    ) -> miette::Result<EValue> {
        match self {
            EDataType::Boolean => json_expected(data.as_bool(), data, "bool").map(EValue::from),
            EDataType::Number => json_expected(data.as_number(), data, "number")
                .map(|num| OrderedFloat(num.as_f64().unwrap()).into()),
            EDataType::String => {
                json_expected(data.as_str(), data, "string").map(|s| s.to_string().into())
            }
            EDataType::Object { ident } => {
                let obj = registry
                    .get_object(ident)
                    .ok_or_else(|| miette!("object id was not present in registry: `{}`", ident))?;

                obj.parse_json(registry, data, inline)
            }
            EDataType::Const { value } => {
                let m = value.matches_json(data);

                if !m.by_type {
                    bail!(
                        "invalid data type. Expected {} but got {}",
                        value,
                        json_kind(data)
                    )
                }

                if !m.by_value {
                    bail!("invalid constant. Expected {} but got {}", value, data)
                }

                Ok(value.default_value())
            }
            EDataType::List { id } => {
                let list = registry.get_list(id).ok_or_else(|| {
                    miette!(
                        "!!INTERNAL ERROR!! list id was not present in registry: `{}`",
                        id
                    )
                })?;

                let JsonValue::Array(items) = data else {
                    bail!(
                        "invalid data type. Expected list but got {}",
                        json_kind(data)
                    )
                };

                let mut list_items = vec![];
                for (i, x) in items.iter_mut().enumerate() {
                    list_items.push(
                        list.value_type
                            .parse_json(registry, x, false)
                            .with_context(|| format!("at index {}", i))?,
                    );
                }

                Ok(EValue::List {
                    id: *id,
                    values: list_items,
                })
            }
            EDataType::Map { id } => {
                let map = registry.get_map(id).ok_or_else(|| {
                    miette!(
                        "!!INTERNAL ERROR!! map id was not present in registry: `{}`",
                        id
                    )
                })?;

                let JsonValue::Object(obj) = data else {
                    bail!(
                        "invalid data type. Expected map but got {}",
                        json_kind(data)
                    )
                };

                let mut entries = BTreeMap::new();

                for (k, v) in obj {
                    let key_name = k.clone();
                    let (k, v) = m_try(|| {
                        let k = map.key_type.parse_json(
                            registry,
                            &mut JsonValue::String(k.clone()),
                            false,
                        )?;
                        let v = map.value_type.parse_json(registry, v, false)?;
                        Ok((k, v))
                    })
                    .with_context(|| format!("in entry with key `{}`", key_name))?;

                    entries.insert(k, v);
                }

                Ok(EValue::Map {
                    id: *id,
                    values: entries,
                })
            }
            EDataType::Unknown => {
                bail!("cannot parse unknown type")
            }
        }
    }
}
