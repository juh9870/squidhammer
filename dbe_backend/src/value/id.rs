use crate::project::module::DbeModule;
use crate::value::id::editor_id::EditorId;
use std::path::Path;

pub mod editor_id;
pub mod parsing;

macro_rules! id_type {
    ($ident:ident) => {
        #[derive(
            Copy, Clone, serde::Serialize, serde::Deserialize, Eq, PartialEq, Hash, Ord, PartialOrd,
        )]
        #[serde(transparent)]
        pub struct $ident(EditorId);

        impl $ident {
            pub fn parse(data: &str) -> miette::Result<Self> {
                Ok(Self(EditorId::parse(data)?))
            }

            pub(crate) fn from_raw(raw: ustr::Ustr) -> $ident {
                Self(EditorId(raw))
            }

            pub fn as_raw(&self) -> &str {
                self.0 .0.as_str()
            }
        }

        impl std::fmt::Display for $ident {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.0)
            }
        }

        impl std::fmt::Debug for $ident {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}({})", stringify!($ident), self.0)
            }
        }

        impl std::str::FromStr for $ident {
            type Err = miette::Error;

            fn from_str(s: &str) -> Result<Self, Self::Err> {
                $ident::parse(s)
            }
        }
    };
}

id_type!(ETypeId);

impl ETypeId {
    pub fn from_path(module: &DbeModule, path: impl AsRef<Path>) -> miette::Result<Self> {
        Ok(Self(EditorId::from_path(module, path.as_ref())?))
    }

    pub fn strip_generics(&self) -> Self {
        let Some(id) = self.0 .0.split('<').next() else {
            return *self;
        };
        Self(EditorId(id.into()))
    }

    pub fn has_generics(&self) -> bool {
        self.0 .0.contains('<')
    }
}

// id_type!(EValueIdStr);

id_type!(EListId);

id_type!(EMapId);

// #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Deserialize)]
// #[serde(untagged)]
// pub enum EValueId {
//     String(EValueIdStr),
//     Numeric(i32),
// }

// impl EValueId {
//     pub fn parse_json(json: &mut JsonValue) -> miette::Result<EValueId> {
//         match json {
//             Value::Number(num) => {
//                 let num = num.as_f64().unwrap();
//                 if num < 0.0 {
//                     bail!("negative numeric ID: {}", num)
//                 }
//                 if num > i32::MAX as f64 {
//                     bail!(
//                         "numeric ID too large, must be at most {}, but got {}",
//                         i32::MAX,
//                         num
//                     )
//                 }
//                 Ok(EValueId::Numeric(num as i32))
//             }
//             Value::String(str) => Ok(EValueId::String(EValueIdStr::parse(str)?)),
//             other => {
//                 bail!(
//                     "invalid data type. Expected string or number but got {}",
//                     json_kind(other)
//                 )
//             }
//         }
//     }
// }
//
// impl Display for EValueId {
//     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
//         match self {
//             EValueId::String(str) => str.fmt(f),
//             EValueId::Numeric(num) => num.fmt(f),
//         }
//     }
// }

#[cfg(test)]
mod tests {
    use super::ETypeId;
    use rstest::rstest;

    #[rstest]
    #[case("namespace:id")]
    #[case("namespace_123:a1/a2/a3/4/5/6")]
    #[case("eh:objects/faction")]
    fn should_parse_type_id(#[case] id: &str) {
        assert!(ETypeId::parse(id).is_ok());
    }

    #[test]
    fn should_fail_empty() {
        assert!(ETypeId::parse("").is_err());
    }

    #[test]
    fn should_fail_no_colon() {
        assert!(ETypeId::parse("some_name").is_err());
    }

    #[test]
    fn should_fail_empty_namespace() {
        assert!(ETypeId::parse(":some_name").is_err());
    }

    #[test]
    fn should_fail_empty_path() {
        assert!(ETypeId::parse("some_name:").is_err());
    }

    #[test]
    fn should_fail_slashes_in_namespace() {
        assert!(ETypeId::parse("namespace/other:path").is_err());
    }

    #[test]
    fn should_fail_capitalized() {
        assert!(ETypeId::parse("namespace/Path").is_err());
        assert!(ETypeId::parse("Namespace/path").is_err());
    }

    #[rstest]
    #[case("name space:id")]
    #[case("namespace:path.other")]
    #[case("namespace:path-other")]
    fn should_fail_invalid_characters(#[case] id: &str) {
        assert!(ETypeId::parse(id).is_err());
    }
}
