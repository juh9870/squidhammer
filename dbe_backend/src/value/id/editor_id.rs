use std::fmt::{Debug, Display, Formatter};
use std::path::Path;
use std::str::FromStr;

use crate::project::module::DbeModule;
use crate::project::TYPES_FOLDER;
use crate::value::id::parsing::normalize_id;
use itertools::Itertools;
use miette::{bail, miette};
use serde::{Deserializer, Serializer};
use ustr::Ustr;

#[inline(always)]
pub fn bad_namespace_char(c: char) -> bool {
    !matches!(c, 'a'..='z' | '0'..='9' | '_')
}

#[inline(always)]
pub fn bad_path_char(c: char) -> bool {
    !matches!(c, 'a'..='z' | '0'..='9' | '_' | '/')
}

pub fn namespace_errors(namespace: &str) -> Option<(usize, char)> {
    namespace.chars().find_position(|c| bad_namespace_char(*c))
}

pub fn path_errors(namespace: &str) -> Option<(usize, char)> {
    namespace.chars().find_position(|c| bad_path_char(*c))
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
pub struct EditorId(pub Ustr);

impl serde::Serialize for EditorId {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.0.serialize(serializer)
    }
}

struct EditorIdVisitor;

impl serde::de::Visitor<'_> for EditorIdVisitor {
    type Value = EditorId;

    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
        formatter.write_str("a string")
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        let id = normalize_id(v);
        Ok(EditorId(id.into()))
    }
}

impl<'de> serde::Deserialize<'de> for EditorId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_string(EditorIdVisitor)
    }
}

impl EditorId {
    pub fn parse(data: &str) -> miette::Result<Self> {
        let (namespace, path): (&str, &str) = data
            .split(':')
            .collect_tuple()
            .ok_or_else(|| miette!("Type path must be in a form of `namespace:path`"))?;

        let namespace = Namespace::from_str(namespace)?;

        if path.is_empty() {
            bail!("Path can't be empty")
        }

        if let Some((i, c)) = path_errors(path) {
            bail!(
                "Invalid symbol `{c}` in path, at position {}",
                i + namespace.id.len() + 1
            )
        }

        Ok(EditorId(data.into()))
    }

    pub fn from_path(module: &DbeModule, path: &Path) -> miette::Result<Self> {
        let sub_path = path
            .strip_prefix(module.path.join(TYPES_FOLDER))
            .map_err(|_| {
                miette!(
                    "Type is outside of types root folder.\nType: `{}`",
                    path.display()
                )
            })?
            .components()
            .collect_vec();
        if sub_path.is_empty() {
            bail!("Malformed type path: `{}`", path.display())
        }

        let namespace = &module.namespace;

        let segments: Vec<String> = sub_path
            .into_iter()
            .with_position()
            .map(|(pos, path)| {
                let str = if matches!(pos, itertools::Position::Last | itertools::Position::Only) {
                    let p: &Path = path.as_ref();
                    p.file_stem().ok_or_else(||miette!("Final path segment has an empty filename"))?.to_string_lossy().to_string()
                } else {
                    path.as_os_str().to_string_lossy().to_string()
                };
                if let Some((i, c)) = path_errors(&str) {
                    bail!("Path folder or file contains invalid symbol `{c}` at position {i} in segment `{}`", path.as_os_str().to_string_lossy().to_string())
                }

                Ok(str)
            })
            .try_collect()?;

        let path = segments.join("/");

        if path.is_empty() {
            bail!("Type can't be placed in a root of types folder")
        }

        Self::parse(&format!("{namespace}:{path}"))
    }
    // #[inline(always)]
    // pub fn raw(&self) -> &Ustr {
    //     &self.0
    // }
}

impl FromStr for EditorId {
    type Err = miette::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        EditorId::parse(s)
    }
}

impl Display for EditorId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            EditorId(id) => write!(f, "{}", id),
        }
    }
}

#[derive(Clone, Eq, PartialEq, Hash)]
pub struct Namespace {
    id: String,
}

impl Debug for Namespace {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        Debug::fmt(&self.id, f)
    }
}

impl Display for Namespace {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.id, f)
    }
}

impl AsRef<str> for Namespace {
    fn as_ref(&self) -> &str {
        &self.id
    }
}

impl From<Namespace> for String {
    fn from(value: Namespace) -> Self {
        value.id
    }
}

impl FromStr for Namespace {
    type Err = miette::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if s.is_empty() {
            bail!("Namespace can't be empty")
        }

        if let Some((i, c)) = namespace_errors(s) {
            bail!("Invalid symbol `{c}` in namespace, at position {i}")
        }

        Ok(Namespace { id: s.into() })
    }
}

struct NamespaceVisitor;

impl serde::de::Visitor<'_> for NamespaceVisitor {
    type Value = Namespace;

    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
        formatter.write_str("a string")
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        match Namespace::from_str(v) {
            Err(e) => Err(serde::de::Error::custom(e)),
            Ok(v) => Ok(v),
        }
    }
}

impl<'de> serde::Deserialize<'de> for Namespace {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_string(NamespaceVisitor)
    }
}
