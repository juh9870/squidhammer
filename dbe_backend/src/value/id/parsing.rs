use crate::etype::econst::ETypeConst;
use crate::etype::EDataType;
use crate::value::id::editor_id::EditorId;
use crate::value::id::ETypeId;
use logos::{Lexer, Logos};
use miette::Diagnostic;
use thiserror::Error;

#[derive(Debug)]
pub enum ParsedId {
    Simple(EDataType),
    GenericObject {
        ident: ETypeId,
        generics: Vec<(String, ParsedId)>,
    },
    List(Box<ParsedId>),
    Map {
        key: Box<ParsedId>,
        value: Box<ParsedId>,
    },
}

pub fn parse_full_id(id: &str) -> Result<ParsedId, IdParseError> {
    let mut lex = PeekableLexer::new(Token::lexer(id));
    let ty = parse_type(&mut lex)?;

    match lex.next() {
        None => Ok(ty),
        Some(Err(())) => Err(IdParseError::BadToken(lex.slice().to_string())),
        Some(Ok(_)) => Err(IdParseError::UnexpectedToken(
            lex.slice().to_string(),
            "end of input",
        )),
    }
}

/// Normalizes an identifier, stripping redundant whitespaces
pub fn normalize_id(id: &str) -> String {
    let mut lex = Token::lexer(id);
    let mut result = String::new();
    while lex.next().is_some() {
        result.push_str(lex.slice());
    }

    result
}

/// Escapes a string for the standard display
pub fn escape_const_string(str: &str) -> String {
    str.replace("'", "''")
}

pub fn unescape_const_string(str: &str) -> String {
    let str = str.strip_prefix("'").unwrap().strip_suffix("'").unwrap();
    str.replace("''", "'")
}

#[derive(Debug, Clone, Error, Diagnostic)]
pub enum IdParseError {
    #[error("unexpected end of input")]
    UnexpectedEnd,
    #[error("unexpected token {0}, expected {1}")]
    UnexpectedToken(String, &'static str),
    #[error("invalid input: `{0}`")]
    BadToken(String),
}

#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n\f]+")]
enum Token {
    #[regex("[a-zA-Z0-9_]+:[a-zA-Z0-9/_]+", |lex| EditorId::parse(lex.slice()).unwrap())]
    Id(EditorId),
    #[token("number")]
    Number,
    #[token("string")]
    String,
    #[token("boolean")]
    Boolean,
    #[token("unknown")]
    Unknown,
    #[token("null")]
    Null,
    #[token("List")]
    List,
    #[token("Map")]
    Map,
    #[token("Item")]
    Item,
    #[token("Key")]
    Key,
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*")]
    OtherAlphanumeric,
    #[regex(r"[-+]?(?:0|[1-9]\d*)(?:\.\d*)?(?:[eE][-+]?\d+)?", |lex| lex.slice().parse::<f64>().unwrap()
    )]
    ConstNumber(f64),
    #[regex(r"'(?:[^']*(?:''[^']*)*)'", |lex| unescape_const_string(lex.slice()))]
    ConstString(String),
    #[regex(r"true|false", |lex| lex.slice().parse::<bool>().unwrap())]
    ConstBool(bool),
    #[token("<")]
    OpenAngle,
    #[token(">")]
    CloseAngle,
    #[token("=")]
    Equals,
    #[token(",")]
    Comma,
}

impl Token {
    fn expected_str(&self) -> &'static str {
        match self {
            Token::Id(_) => "identifier",
            Token::Number => "number",
            Token::String => "string",
            Token::Boolean => "boolean",
            Token::Unknown => "unknown",
            Token::Null => "null",
            Token::List => "List",
            Token::Map => "Map",
            Token::ConstNumber(_) => "constant number",
            Token::ConstString(_) => "constant string",
            Token::ConstBool(_) => "constant boolean",
            Token::OpenAngle => "`<` symbol",
            Token::CloseAngle => "`>` symbol",
            Token::Equals => "`=` symbol",
            Token::Comma => "`,` symbol",
            Token::OtherAlphanumeric => "alphanumeric identifier",
            Token::Item => "`Item` identifier",
            Token::Key => "`Key` identifier",
        }
    }

    fn is_alphanumeric(&self) -> bool {
        matches!(
            self,
            Token::Number
                | Token::String
                | Token::Boolean
                | Token::Unknown
                | Token::Null
                | Token::List
                | Token::Map
                | Token::Item
                | Token::Key
                | Token::OtherAlphanumeric
        )
    }

    fn expect_same(&self, self_slice: &str, expected: Token) -> Result<(), IdParseError> {
        if self == &expected {
            Ok(())
        } else {
            Err(IdParseError::UnexpectedToken(
                self_slice.to_string(),
                expected.expected_str(),
            ))
        }
    }
}

fn parse_type(tokens: &mut PeekableLexer) -> Result<ParsedId, IdParseError> {
    fn next_token_opt(tokens: &mut PeekableLexer) -> Result<Option<Token>, IdParseError> {
        tokens
            .next()
            .transpose()
            .map_err(|()| IdParseError::BadToken(tokens.slice().to_string()))
    }

    fn peek_token_opt<'a>(
        tokens: &'a mut PeekableLexer<'_>,
    ) -> Result<Option<&'a Token>, IdParseError> {
        let slice = tokens.peek_slice();
        match tokens.peek() {
            None => Ok(None),
            Some(Err(())) => Err(IdParseError::BadToken(slice.to_string())),
            Some(Ok(token)) => Ok(Some(token)),
        }
    }

    fn next_token(tokens: &mut PeekableLexer) -> Result<Token, IdParseError> {
        next_token_opt(tokens)?.ok_or(IdParseError::UnexpectedEnd)
    }

    fn expect_next(tokens: &mut PeekableLexer, expected: Token) -> Result<(), IdParseError> {
        next_token(tokens)?.expect_same(tokens.slice(), expected)
    }

    fn parse_generics(tokens: &mut PeekableLexer) -> Result<Vec<(String, ParsedId)>, IdParseError> {
        let mut generics = Vec::new();
        let next = peek_token_opt(tokens)?;
        match next {
            None => {
                return Ok(generics);
            }
            Some(token) => {
                match token {
                    Token::CloseAngle | Token::Comma => {
                        return Ok(generics);
                    }
                    Token::OpenAngle => {
                        // continue
                    }
                    _ => {
                        return Err(IdParseError::UnexpectedToken(
                            tokens.peek_slice().to_string(),
                            "generic arguments",
                        ));
                    }
                }
            }
        }
        loop {
            let ident = next_token(tokens)?;
            if !ident.is_alphanumeric() {
                return Err(IdParseError::UnexpectedToken(
                    tokens.slice().to_string(),
                    "alphanumeric identifier",
                ));
            }
            let identifier = tokens.slice();

            expect_next(tokens, Token::Equals)?;

            let ty = parse_type(tokens)?;
            generics.push((identifier.to_string(), ty));

            let next = next_token(tokens)?;

            match next {
                Token::CloseAngle => {
                    return Ok(generics);
                }
                Token::Comma => {
                    // continue
                }
                _ => {
                    return Err(IdParseError::UnexpectedToken(
                        tokens.slice().to_string(),
                        "`,` or `>`",
                    ))
                }
            }
        }
    }

    match next_token(tokens)? {
        Token::Id(id) => {
            let generics = parse_generics(tokens)?;

            if generics.is_empty() {
                return Ok(ParsedId::Simple(EDataType::Object { ident: ETypeId(id) }));
            }

            Ok(ParsedId::GenericObject {
                ident: ETypeId(id),
                generics,
            })
        }
        Token::Map => {
            expect_next(tokens, Token::OpenAngle)?;
            expect_next(tokens, Token::Key)?;
            expect_next(tokens, Token::Equals)?;
            let key = Box::new(parse_type(tokens)?);
            expect_next(tokens, Token::Comma)?;
            expect_next(tokens, Token::Item)?;
            expect_next(tokens, Token::Equals)?;
            let value = Box::new(parse_type(tokens)?);
            expect_next(tokens, Token::CloseAngle)?;
            Ok(ParsedId::Map { key, value })
        }
        Token::List => {
            expect_next(tokens, Token::OpenAngle)?;
            expect_next(tokens, Token::Item)?;
            expect_next(tokens, Token::Equals)?;
            let ty = Box::new(parse_type(tokens)?);
            expect_next(tokens, Token::CloseAngle)?;
            Ok(ParsedId::List(ty))
        }
        Token::Number => Ok(ParsedId::Simple(EDataType::Number)),
        Token::String => Ok(ParsedId::Simple(EDataType::String)),
        Token::Boolean => Ok(ParsedId::Simple(EDataType::Boolean)),
        Token::Unknown => Ok(ParsedId::Simple(EDataType::Unknown)),
        Token::Null => Ok(ParsedId::Simple(EDataType::null())),
        Token::ConstNumber(num) => Ok(ParsedId::Simple(EDataType::Const { value: num.into() })),
        Token::ConstString(value) => Ok(ParsedId::Simple(EDataType::Const {
            value: ETypeConst::String(value.into()),
        })),
        Token::ConstBool(value) => Ok(ParsedId::Simple(EDataType::Const {
            value: value.into(),
        })),
        Token::OpenAngle
        | Token::CloseAngle
        | Token::Equals
        | Token::Comma
        | Token::Item
        | Token::Key
        | Token::OtherAlphanumeric => Err(IdParseError::UnexpectedToken(
            tokens.slice().to_string(),
            "type or id",
        )),
    }
}

struct PeekableLexer<'a> {
    lexer: Lexer<'a, Token>,
    peek_state: Option<PeekState<'a>>,
}

struct PeekState<'a> {
    next: Option<Result<Token, ()>>,
    cur_slice: &'a str,
}

impl<'a> PeekableLexer<'a> {
    pub fn new(lexer: Lexer<'a, Token>) -> Self {
        Self {
            lexer,
            peek_state: None,
        }
    }

    pub fn slice(&mut self) -> &'a str {
        if let Some(state) = self.peek_state.as_mut() {
            state.cur_slice
        } else {
            self.lexer.slice()
        }
    }

    pub fn peek(&mut self) -> Option<&Result<Token, ()>> {
        self.ensure_peeked();

        self.peek_state.as_ref().unwrap().next.as_ref()
    }

    pub fn peek_slice(&mut self) -> &'a str {
        self.ensure_peeked();

        self.lexer.slice()
    }

    fn ensure_peeked(&mut self) {
        if self.peek_state.is_none() {
            self.peek_state = Some(PeekState {
                cur_slice: self.lexer.slice(),
                next: self.lexer.next(),
            });
        }
    }
}

impl<'a> Iterator for PeekableLexer<'a> {
    type Item = Result<Token, ()>;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(state) = self.peek_state.take() {
            return state.next;
        }
        self.lexer.next()
    }
}
