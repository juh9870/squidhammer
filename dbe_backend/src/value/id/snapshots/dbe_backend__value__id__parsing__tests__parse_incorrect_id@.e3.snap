---
source: dbe_backend/src/value/id/parsing.rs
description: "For Id: .e3"
---
invalid input: `.`
